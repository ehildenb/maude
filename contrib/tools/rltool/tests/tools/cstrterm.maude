load ../../systems/choice.maude
load ../../rltool.maude

mod RLTOOL-TEST is
  pr RLTOOL .

  var X Y Z : Variable .

  ops ct : Variable Variable Variable -> CTerm .
  eq ct(X,Y,Z) = ('`{_`}['__[X,Y]] | '_=C_[X,Z] ?= 'tt.Pred) .
endm

set show timing off .
set show stats off .
set show advisories off .

print reveal _;_ __ _,_ _&_ .

red ct('X:Nat,'Y:Nat,'Z:Nat) :: CTerm .

--- check term above is well-formeed in module
red wellFormed(['CHOICE-DATA],ct('X:Nat,'Y:Nat,'Z:Nat)) == true .

--- check that this term is ill-formed
red wellFormed(['CHOICE-DATA],ct('X:Nat,'Y:Foo,'Z:Nat)) == false .

--- check condition replacement/conjoining/negation/NNF works
red setCond(ct('X:Nat,'Y:Nat,'Z:Nat),'_=C_['X:Nat,'N:Nat] ?= 'tt.Pred) .
red conj(ct('X:Nat,'Y:Nat,'Z:Nat),'_=C_['X:Nat,'N:Nat] ?= 'tt.Pred) .
red negateCond(conj(ct('X:Nat,'Y:Nat,'Z:Nat),'_=C_['X:Nat,'N:Nat] ?= 'tt.Pred)) .
red toNNF(negateCond(conj(ct('X:Nat,'Y:Nat,'Z:Nat),'_=C_['X:Nat,'N:Nat] ?= 'tt.Pred))) .

--- check that variable extraction works (also for sets)
red vars(noterm) .
red termVars(noterm) .
red formVars(noterm) .

red vars(ct('X:Nat,'Y:Nat,'Z:Nat)) .
red termVars(ct('X:Nat,'Y:Nat,'Z:Nat)) .
red formVars(ct('X:Nat,'Y:Nat,'Z:Nat)) .

red vars(ct('X:Nat,'Y:Nat,'Z:Nat) | ct('N:Nat,'V:Nat,'Q:Nat)) .
red termVars(ct('X:Nat,'Y:Nat,'Z:Nat) | ct('N:Nat,'V:Nat,'Q:Nat)) .
red formVars(ct('X:Nat,'Y:Nat,'Z:Nat) | ct('N:Nat,'V:Nat,'Q:Nat)) .

--- check shared variable definitions (no shared vars)
red sharedVars(noterm) .
red sharedVars(ct('X:Nat,'Y:Nat,'Z:Nat) | ct('N:Nat,'V:Nat,'Q:Nat)) .

--- check shared variables definitions (with shared vars)
red sharedVars(ct('X:Nat,'Y:Nat,'Z:Nat) | ct('N:Nat,'V:Nat,'Y:Nat)) .

--- check shared variables BETWEEN two term sets
red sharedVars(ct('X:Nat,'Y:Nat,'Z:Nat) | ct('N:Nat,'V:Nat,'Y:Nat),ct('W:Nat,'Y:Nat,'Q:Nat) | ct('X:Nat,'V:Nat,'Y:Nat))  .

quit
