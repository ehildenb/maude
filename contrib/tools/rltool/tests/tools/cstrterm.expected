
	    Maude Reachability Logic Prover v0.2.9 -- Sep. 21st 2018

==========================================
reduce in RLTOOL-TEST : ct('X:Nat, 'Y:Nat, 'Z:Nat) :: CTerm .
result Bool: true
==========================================
reduce in RLTOOL-TEST : wellFormed(['CHOICE-DATA], ct('X:Nat, 'Y:Nat, 'Z:Nat)) == true .
result Bool: true
==========================================
reduce in RLTOOL-TEST : wellFormed(['CHOICE-DATA], ct('X:Nat, 'Y:Foo, 'Z:Nat)) == false .
result Bool: true
==========================================
reduce in RLTOOL-TEST : setCond(ct('X:Nat, 'Y:Nat, 'Z:Nat), 'tt.Pred ?= '_=C_['X:Nat,'N:Nat]) .
result CCTerm: ('`{_`}['__['X:Nat,'Y:Nat]] | 'tt.Pred ?= '_=C_['X:Nat,'N:Nat])
==========================================
reduce in RLTOOL-TEST : conj(ct('X:Nat, 'Y:Nat, 'Z:Nat), 'tt.Pred ?= '_=C_['X:Nat,'N:Nat]) .
result CCTerm: ('`{_`}['__['X:Nat,'Y:Nat]] | 'tt.Pred ?= '_=C_['X:Nat,'N:Nat] /\ 'tt.Pred ?= '_=C_['X:Nat,'Z:Nat])
==========================================
reduce in RLTOOL-TEST : negateCond(conj(ct('X:Nat, 'Y:Nat, 'Z:Nat), 'tt.Pred ?= '_=C_['X:Nat,'N:Nat])) .
result QFCTerm: ('`{_`}['__['X:Nat,'Y:Nat]] | ~ ('tt.Pred ?= '_=C_['X:Nat,'N:Nat] /\ 'tt.Pred ?= '_=C_['X:Nat,'Z:Nat]))
==========================================
reduce in RLTOOL-TEST : toNNF(negateCond(conj(ct('X:Nat, 'Y:Nat, 'Z:Nat), 'tt.Pred ?= '_=C_['X:Nat,'N:Nat]))) .
result QFCTerm: ('`{_`}['__['X:Nat,'Y:Nat]] | 'tt.Pred != '_=C_['X:Nat,'N:Nat] \/ 'tt.Pred != '_=C_['X:Nat,'Z:Nat])
==========================================
reduce in RLTOOL-TEST : vars(noterm) .
result EmptyTypeSet: (none).EmptyTypeSet
==========================================
reduce in RLTOOL-TEST : termVars(noterm) .
result EmptyTypeSet: (none).EmptyTypeSet
==========================================
reduce in RLTOOL-TEST : formVars(noterm) .
result EmptyTypeSet: (none).EmptyTypeSet
==========================================
reduce in RLTOOL-TEST : vars(ct('X:Nat, 'Y:Nat, 'Z:Nat)) .
result NeVariableSet: 'X:Nat ; 'Y:Nat ; 'Z:Nat
==========================================
reduce in RLTOOL-TEST : termVars(ct('X:Nat, 'Y:Nat, 'Z:Nat)) .
result NeVariableSet: 'X:Nat ; 'Y:Nat
==========================================
reduce in RLTOOL-TEST : formVars(ct('X:Nat, 'Y:Nat, 'Z:Nat)) .
result NeVariableSet: 'X:Nat ; 'Z:Nat
==========================================
reduce in RLTOOL-TEST : vars(ct('N:Nat, 'V:Nat, 'Q:Nat) | ct('X:Nat, 'Y:Nat, 'Z:Nat)) .
result NeVariableSet: 'N:Nat ; 'Q:Nat ; 'V:Nat ; 'X:Nat ; 'Y:Nat ; 'Z:Nat
==========================================
reduce in RLTOOL-TEST : termVars(ct('N:Nat, 'V:Nat, 'Q:Nat) | ct('X:Nat, 'Y:Nat, 'Z:Nat)) .
result NeVariableSet: 'N:Nat ; 'V:Nat ; 'X:Nat ; 'Y:Nat
==========================================
reduce in RLTOOL-TEST : formVars(ct('N:Nat, 'V:Nat, 'Q:Nat) | ct('X:Nat, 'Y:Nat, 'Z:Nat)) .
result NeVariableSet: 'N:Nat ; 'Q:Nat ; 'X:Nat ; 'Z:Nat
==========================================
reduce in RLTOOL-TEST : sharedVars(noterm) .
result EmptyTypeSet: (none).EmptyTypeSet
==========================================
reduce in RLTOOL-TEST : sharedVars(ct('N:Nat, 'V:Nat, 'Q:Nat) | ct('X:Nat, 'Y:Nat, 'Z:Nat)) .
result EmptyTypeSet: (none).EmptyTypeSet
==========================================
reduce in RLTOOL-TEST : sharedVars(ct('N:Nat, 'V:Nat, 'Y:Nat) | ct('X:Nat, 'Y:Nat, 'Z:Nat)) .
result Variable: 'Y:Nat
==========================================
reduce in RLTOOL-TEST : sharedVars(ct('N:Nat, 'V:Nat, 'Y:Nat) | ct('X:Nat, 'Y:Nat, 'Z:Nat), ct('W:Nat, 'Y:Nat, 'Q:Nat) | ct('X:Nat, 'V:Nat, 'Y:Nat)) .
result NeVariableSet: 'V:Nat ; 'X:Nat ; 'Y:Nat
Bye.
